### PR TITLE
feat(mobile-screenshots): probe one iOS shard first and fan out only when it changed

### PR DESCRIPTION
**Pull request by marcodejongh**
## Summary

Fourth PR of the screenshot pipeline (after #5320, #5324, #5325). The iOS App Store capture becomes self-gating: one probe shard (en-US × iPhone 16 Pro Max) is captured first and compared against the last full set; the other 11 locale × device shards run only when it changed. An unchanged run costs one macOS shard instead of twelve (the public-repo cap is 5 concurrent macOS jobs, so a full run is three waves).

- `scripts/compare-screenshots.ts` (`vp run screenshot:compare`): pairs PNGs by name, flags missing / extra / resized, and does a raw RGB compare with a per-channel tolerance (8) and a max differing-pixel ratio (0.1%); writes a summary JSON, `changed=` / `changed_files=` outputs, a step-summary table and red-mask diff PNGs. The header documents how to recalibrate the thresholds from two identical captures.
- `scripts/screenshot-baseline.ts` (`vp run screenshot:baseline`): `pack` / `publish` / `fetch` against a rolling `screenshots-baseline` prerelease (one zip per locale × device plus a manifest with per-file sha256 and the source commit, uploaded last). iOS PNGs stay out of git, as decided in #2905.
- `.github/actions/ios-screenshot-shard`: the per-shard capture body (cache restore, pinned Maestro, two-attempt capture, debug + shard artifacts) shared by the probe and the fan-out.
- `mobile-screenshots-ios.yml`: `gate` (`full` | `probe`) and `publish_baseline` inputs; `setup` emits the matrix JSON (probe pair excluded) and the gate; new `probe` job; `ios-capture` runs only when the probe changed (or was skipped) with `max-parallel: 4`; `ios-finalize` succeeds on both paths, publishes the baseline after a green full run (automatic runs, or `publish_baseline: true`), and uploads to App Store Connect on the automatic trigger (a later PR adds that trigger) or on `upload: true` as before.
- `mobile-store-draft.yml`: after the draft version is created, fetch the baseline and run `fastlane ios screenshots` (best-effort), so a draft created after the capture finished still gets the screenshots.
- Android is untouched here; its compare step lands after #5320.

Author-side verification: scripts project 1971 tests (54 new); mutation kills for the ratio boundary, the any/all aggregation and a dropped locale; a real `pack` produced 15 flat zips + manifest; `vp run typecheck:scripts` clean; all workflow YAML parses. Live verification on this branch, all three runs green:
- [run 34216902625](https://github.com/boardsesh/boardsesh/actions/runs/34216902625) `gate=probe`, no baseline yet: the probe reported changed, the matrix exclude left exactly the 11 other shards, and `ios-finalize` ran with the empty `environment:` expression.
- [run 34229298504](https://github.com/boardsesh/boardsesh/actions/runs/34229298504) `gate=probe publish_baseline=true`: full set captured, dimension gate green, and the `screenshots-baseline` prerelease now holds the 15 shard zips plus `ios-manifest.json`.
- [run 34237649574](https://github.com/boardsesh/boardsesh/actions/runs/34237649574) `gate=probe` against that baseline: the pixel compare ran for real — 10 shots, 8 byte-identical, 2 changed (`02-home.png`: live feed content; `05-workout-generator.png`: 0.32% of pixels), so it fanned out. Both differences are live data, which #5330 (replay + frozen clock) removes; the 8 identical shots show the runner renders deterministically, so the thresholds hold.

Follow-ups tracked for this PR: rebase once #5320 lands (it removes the `schedule:` that is still here), expose the `fixtures` input through the composite action once #5330 lands, and the Android compare + auto-commit step.

## Release Notes

- [x] No release note needed (internal / technical change)

## Test plan

1. CI green.

## Risk

Risk: 3/5 — CI workflow surgery and a publish step; no app code

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01VjmyzX7wp1cgksH1zweWWA

### CODE REVIEW

File: .github/workflows/mobile-screenshots-ios.yml [hunk 1]
```diff
@@ -225,164 +302,153 @@ jobs:
...
-        uses: actions/upload-artifact@v4
+  ios-capture:
+    needs: [setup, ios-build, probe]
+    # `probe` is skipped on a full run and must not skip the fan-out with it, so
+    # this is written on `result` rather than the default all-needs-succeeded
+    # gate. `!cancelled()` keeps a cancelled run from starting 11 macOS shards.
+    if: >-
+      !cancelled() && needs.setup.result == 'success' && needs.ios-build.result == 'success' &&
+      (needs.probe.result == 'skipped' ||
+       (needs.probe.result == 'success' && needs.probe.outputs.changed == 'true'))
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply .github/workflows/mobile-screenshots-ios.yml]: Fixed in ac568a24c9 (`scripts/screenshot-probe-scope.ts`): the probe job now runs `git diff --name-only <baseline_commit> <source_sha>` after the baseline fetch (fetching the baseline commit explicitly first, and treating an unreachable commit as force-full) and calls `vp run screenshot:probe-scope` against the changed paths. It forces `force_full=true` when a change matches one of five documented rules (non-en-US `packages/shared/i18n/locales/**`, `packages/mobile/locales/**`, an ipad/tablet-matching path under `packages/mobile/`, `packages/mobile/app.config.ts`, or `app-stores/apple/**`), each with its own unit test plus tests confirming an en-US-only locale change and an unrelated file do NOT force. `ios-capture` and `ios-finalize` (the `UNCHANGED" env) now treat `force_full == '"true"'` exactly like `changed == '"true"'`, and the step summary explains why. Contract coverage: `scripts/__tests__/mobile-screenshots-workflows.test.ts` pins that the probe calls the scope script and that the capture condition references `force_full`.

File: scripts/screenshot-baseline.ts [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+    mkdirSync(target, { recursive: true });
+    runUnzip(runner, ['-o', '-q', join(downloadDir, zipName), '-d', target]);
+    unzipped.push(target);
+
+    if (manifest && shard) {
+      const sha256ByFile: Record<string, string> = {};
+      for (const name of readdirSync(target).filter((entry) => entry.toLowerCase().endsWith('.png'))) {
+        sha256ByFile[name] = sha256Of(join(target, name));
+      }
+      const verification = verifyShardAgainstManifest(manifest, target, shard.locale, shard.deviceSlug, sha256ByFile);
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply scripts/screenshot-baseline.ts]: Fixed in ac568a24c9 (scripts/screenshot-baseline.ts, fetchBaseline / verifyShardAgainstManifest): the manifest is now mandatory for found=true. If ios-manifest.json is missing, unreadable, or fails to parse, fetchBaseline returns found=false with a ::warning:: before touching any zip. Under --all, every (locale, deviceSlug) pair the manifest lists is now required to actually have downloaded; a missing shard zip also fails closed with a ::warning:: naming which asset(s) are short. verifyShardAgainstManifest now checks both directions for a given shard: an extra file present locally but not in the manifest (already covered before), a hash mismatch (already covered), and — new — a file the manifest lists that never landed after unzip.

Tests added/updated in scripts/__tests__/screenshot-baseline.test.ts: manifest missing entirely -> found=false; one asset missing under --all -> found=false; a file the manifest listed that never landed -> flagged (pure-function unit test); an extra untracked file inside a zip -> found=false at the fetchBaseline level (not just the pure-function level); happy path still finds. Three pre-existing tests that had been implicitly relying on "no manifest = skip verification" (unzip exit-code handling, tree-rebuild under --all) were updated to include a valid (empty-files) manifest so they now exercise the new mandatory-manifest path instead of bypassing it.

**Comment by marcodejongh** [comment]
Addressed in 0d3615dd4 (head is now `0d3615dd4`, on top of 17f15b7e1):

1. **Medium — probe failure silently skipping the fan-out.** `ios-capture`'s `if:` now also fans out on `needs.probe.result == 'failure'` (`.github/workflows/mobile-screenshots-ios.yml:413`), matched by a mutation-tested contract test at `scripts/__tests__/mobile-screenshots-workflows.test.ts:145-166` (dropping the `'failure'` clause fails the test; restored and reverified green). `ios-finalize`'s `FULL_SET_CAPTURED` (`.github/workflows/mobile-screenshots-ios.yml:514-517`) no longer requires the probe to have succeeded/skipped — completeness is enforced by the dimension gate behind `success()`, not by the probe's own result — while `UNCHANGED` (`.github/workflows/mobile-screenshots-ios.yml:503-505`) still requires `needs.probe.result == 'success'`, now pinned explicitly at `scripts/__tests__/mobile-screenshots-workflows.test.ts:177-193`. The header's skip trace at `.github/workflows/mobile-screenshots-ios.yml:25-30` and the step summary at `.github/workflows/mobile-screenshots-ios.yml:649-650` ("Probe crashed") were updated to describe the failure path.

2. **Low — release notes never updated after first publish.** `publishBaseline` now runs `gh release edit screenshots-baseline --notes <...>` when the release already exists (`scripts/screenshot-baseline.ts:365-372`), keeping `release create` for first publish only.

3. **Low — no test for the existing-release path.** `scripts/__tests__/screenshot-baseline.test.ts:319-352` now asserts `release create` is never called, `release edit --notes` carries the current commit/run, and `release upload --clobber` runs for both assets with the manifest last.

4. **Low — probe-failure path not asserted.** Covered by item 1's contract test (`scripts/__tests__/mobile-screenshots-workflows.test.ts:145-166`), mutation-verified.

5. **Low — undocumented `--unreachable-baseline` precedence.** Documented in the module comment (`scripts/screenshot-probe-scope.ts:30-34`) and pinned by a new pure `decideProbeScopeFromOptions()` (`scripts/screenshot-probe-scope.ts:128-140`) with tests at `scripts/__tests__/screenshot-probe-scope.test.ts:139-167`, including the both-flags-supplied case.

Also folded in a related gap found while reviewing a sibling PR touching the same probe pipeline: `scripts/compare-screenshots.ts`'s `compareScreenshotSets` had no test for an empty candidate directory (reachable when Maestro crashes before writing any PNG). Verified the existing code already reports `changed: true` with every baseline file `missing` (never `unchanged`) — no code fix needed — and added regression tests for that case plus the symmetric empty-baseline case at `scripts/__tests__/compare-screenshots.test.ts:286-338`.

Validation: `vp check --fix` + `vp check` clean on all touched files; `vp run typecheck:scripts` clean; targeted tests (`mobile-screenshots-workflows`, `screenshot-baseline`, `screenshot-probe-scope`, `compare-screenshots`) all green; full `scripts` project: 114 files / 2310 tests passed; workflow YAML re-parsed successfully after edits.

**Comment by marcodejongh** [comment]
Addressed all three cosmetic notes from the re-review in `f956d4428`:

- **Credential fallback**: both `SCREENSHOT_USER_EMAIL`/`PASSWORD` blocks in `mobile-screenshots-ios.yml` now note the fallback is the documented PROD screenshots account (`docs/mobile-screenshot-fixtures.md`), not a leaked secret, and that #5330 scopes it to the capture step only, outside replay.
- **Diff ratio**: `docs/mobile-store-release.md` now spells out `0.001 (0.001 = 0.1% of the pixels)`.
- **Concurrent publishes**: same doc now notes that two `publish_baseline` dispatches racing can interleave the release-notes edit with the asset uploads and briefly desync which commit the notes name from which shards landed (last writer wins) — acceptable at today's manual, occasional cadence.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01VjmyzX7wp1cgksH1zweWWA

**Comment by marcodejongh** [comment]
Addressed the three items from the latest re-review in \`f6888b966\`:

1. **ASC key guard**: \`mobile-store-draft.yml\`'s screenshot-attach step now checks \`[ -f "$ASC_KEY_PATH" ]\` before running fastlane and fails with a clear \`::error::\` if the key was never decoded. Step stays \`continue-on-error: true\`; the existing "Warn if attaching screenshots failed" step already catches \`outcome == 'failure'\`.
2. **Failed baseline fetch leaves nothing behind**: \`fetchBaseline(--all)\` now wipes the whole \`outDir\` on any verification failure (missing asset, hash mismatch, extra/missing file, bad manifest) via a new \`failFetch()\` helper, not just the one shard that failed. New test: two shards, second fails verification — \`outDir\` no longer exists and \`found\` is false. Mutation-tested (temporarily dropped the \`rmSync\` — exactly that test failed).
3. **Documented the probe's remaining blind spot**: a shared component under \`packages/mobile/src/components/**\` that only renders differently on iPad isn't caught by the pixel probe or the path-based force-full rules, since those match file paths, not rendered output. Documented in both \`docs/mobile-store-release.md\` and the \`screenshot-probe-scope.ts\` module comment, with the intentional remedy (manual \`gate: full\` / \`upload: true\` dispatch) rather than a rule that force-fulls on every component change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01VjmyzX7wp1cgksH1zweWWA

**Comment by marcodejongh** [comment]
On the `capturedAt` question: it is optional on the pack options (`capturedAt?: string`, scripts/screenshot-baseline.ts:291) and defaults to `new Date().toISOString()` when omitted (line 337), so the `refuses to pack a tree that lost a shard` test never writes `capturedAt: undefined` — the manifest always carries a real timestamp. The iPad-specific shared-component blind spot stays as documented; no change.

**Comment by marcodejongh** [comment]
Addressed the review item from the latest #5331 re-review pass, in \`e71e15f34\`:

\`fetchBaseline\` (both \`--all\` and single-asset paths) now skips any downloaded zip whose name \`parseAssetName\` can't decode, or whose (locale, deviceSlug) the manifest's \`files\` doesn't cover — logs a \`::warning::\` naming it and moves on, rather than extracting it (an undecodable name previously dumped straight into \`outDir\`'s root, bypassing verification) or failing the whole fetch over it. A future-format or migration artifact sitting in the release must not brick the probe.

New test: an extra \`ios-legacy.zip\` the manifest doesn't list sits alongside a real shard — it's skipped, and \`found\` stays true once the listed shard verifies. Also updated three pre-existing tests whose empty-manifest fixtures had been relying on the old "verify nothing, extract everything" behavior for unrelated assertions (unzip target paths, exit-code handling) — they now list their shard and match real bytes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01VjmyzX7wp1cgksH1zweWWA

**Comment by marcodejongh** [comment]
Addressed the two real items from the latest re-review, in \`2925923c6\`:

- **Temp directory leaks**: \`fetchBaseline\`'s auto-created download dir and \`publishBaseline\`'s auto-created \`outDir\` (when the caller skips \`--out\`) are now both wrapped in \`try/finally\` and removed on every exit path — success, a \`failFetch()\` early return, or a thrown error. A caller-supplied directory is left alone either way. Manifest-last upload order in \`publishBaseline\` is unchanged. New tests: download dir gone after a successful \`--all\` fetch, both dirs gone after a failed fetch, and the temp \`outDir\` gone after a publish without \`--out\`.
- **\`compare-screenshots.ts\` \`readRawImage\`**: now asserts \`channels === 4\` right after \`ensureAlpha()\` and throws a clear error otherwise, so a short buffer can never silently read as NaN-equals-match in \`findDifferingPixels\`. New test mocks a 3-channel decode and asserts it throws.

The credential-scope item from this same round lands separately in #5330 (fixture record/replay), which scopes \`SCREENSHOT_USER_EMAIL\`/\`PASSWORD\` to the capture step outside replay.

All four fixes mutation-tested; full \`scripts\` project green (114 files / 2316 tests).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01VjmyzX7wp1cgksH1zweWWA

**Comment by marcodejongh** [comment]
Fixed the probe-crash gap in \`78f32f794\`:

\`setup\` now emits two matrices — \`ios_matrix_full\` (exclude: []) and \`ios_matrix_without_probe\` (the probe's own en-US x iPhone 16 Pro Max shard excluded), always computed regardless of \`gate\`. \`ios-capture\`'s \`strategy.matrix\` picks between them on \`needs.probe.result\` at evaluation time — only an actually-successful probe gets the 11-shard exclude; a crashed probe (which never captured or uploaded that shard itself) and a skipped one (gate=full) both get the full 12. \`FULL_SET_CAPTURED\` is unchanged (still just "ios-capture succeeded"), but now that's trustworthy on the probe-failure path since ios-capture itself produces all 12 artifacts there.

Also fixed the comment inaccuracy: the \`FULL_SET_CAPTURED\` rationale comment now names the matrix mechanism instead of vaguely asserting "fans out to every shard."

Contract test pins the matrix expression referencing both outputs and \`needs.probe.result\`; mutation-tested (hardcoding the matrix to always use the excluded one fails it).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01VjmyzX7wp1cgksH1zweWWA
